File: Dynamic Programming 2D/2376: Count Special Integers.js
// https://leetcode.com/problems/count-special-integers/
// Difficulty: Hard
// Tags: dynamic programming 2d, digit dp

// Problem
// We call a positive integer special if all of its digits are distinct.

// Given a positive integer n, return the number of special integers that belong to the interval [1, n].

// Solution, complexities listed inside
// Say we have taken some numbers. For that state, we need to know which numbers we took, so we don't repeat numbers. We also need to know the current index we are inserting a number at. This kind of represents an offset. For instance if our number is at most 5 digits, 00_ _ _ is clearly different than 0 _ _ _ _, even if the digit we insert is the same. We also need to know if we are allowed to break the bounds of the ith digit, this is the tight variable. For instance if n=25, and we pick a 1 for the first digit, we have free choice for the next. If we pick a 2, we are still bounded. We also use the index to compute this.

// Establish a dp state of length(n) * 2 (if tight or not) * bitmask (which numbers we have taken). This is 20*1024. Each state considers up to 10 digits.

// For each state, we check which digits we can take. If we are still tight, we can only take up to the ith digit of the original number. We also cannot take duplicate numbers so we check our mask. We can always take a 0 if we are leading, but once we aren't leading anymore, we can only take a 0 once, and we start marking the mask.

/**
 * @param {number} n
 * @return {number}
 */
var countSpecialNumbers = function (n) {
  const strNum = String(n);
  // memo[index][digitsConsideredMask][tight]
  const memo = new Array(strNum.length)
    .fill()
    .map(() => new Array(1024).fill().map(() => new Array(2).fill(-1)));

  function dp(index, mask, tight) {
    // base case, we are at the end
    if (index === strNum.length) {
      // we don't gain a number if we only took 0s
      return mask === 0 ? 0 : 1;
    }

    if (memo[index][mask][tight] !== -1) {
      return memo[index][mask][tight];
    }

    // what we can go up to if we are tight
    const tightBound = Number(strNum[index]);
    const upperBound = tight ? tightBound : 9;

    let resForThis = 0;

    for (let digit = 0; digit <= upperBound; digit++) {
      // skip numbers we already took
      if ((mask >> digit) & 1) {
        continue;
      }
      // if we were tight, and we took the top digit, we are still tight. if we weren't tight, or we broke the tightness, we arent
      const newTight = tight && digit === Number(strNum[index]) ? 1 : 0;
      let newMask;
      // we can always take a 0 if we are still leading. once we stop leading (our mask is > 0), then we cannot take a 0 anymore
      if (digit === 0) {
        // if we have already taken a number, now we need to mark the 0 as taken, to prevent numbers like 100 from being counted as special
        if (mask > 0) {
          newMask = mask | 1;
        }
        // if we have never taken a number, we should freely be allowed to take 0s without it interfering with duplicate numbers
        else {
          newMask = mask;
        }
      } else {
        newMask = mask | (1 << digit);
      }
      resForThis += dp(index + 1, newMask, newTight);
    }

    memo[index][mask][tight] = resForThis;
    return resForThis;
  }

  return dp(0, 0, 1);
};

File: problems/Leetcode/2376: Count Special Integers.py
# https://leetcode.com/problems/count-special-integers/description/
# Difficulty: Hard
# Tags: digit dp

# Problem
# We call a positive integer special if all of its digits are distinct.

# Given a positive integer n, return the number of special integers that belong to the interval [1, n].

# Solution, O(log n * 2 * 2^(log n)) space, 10 times that for time.
# We need the typical index and tight for digit dp. We also need a mask of which numbers we have taken, we can only take numbers we have not. Which numbers we have taken also informs us if we have taken a non zero which lets us know if we can take a leading 0 or not.
# can maybe compress the for loop, didn't read problem but lots of digit DP can be compressed

class Solution:
    def countSpecialNumbers(self, n: int) -> int:
        strNum = str(n)
        memo = [[[-1 for _ in range(2)] for _ in range(1024)] for _ in range(len(strNum))]

        def dp(index, mask, tight):
            if index == len(strNum):
                # we don't gain a number if we only took 0s
                return 0 if mask == 0 else 1

            if memo[index][mask][tight] != -1:
                return memo[index][mask][tight]

            # what we can go up to if we are tight
            tightBound = int(strNum[index])
            upperBound = tightBound if tight else 9

            resForThis = 0

            for digit in range(upperBound + 1):
                # skip numbers we already took
                if (mask >> digit) & 1:
                    continue

                # if we were tight, and we took the top digit, we are still tight
                # if we weren't tight, or we broke the tightness, we aren't
                newTight = 1 if tight and digit == int(strNum[index]) else 0

                # we can always take a 0 if we are still leading
                # once we stop leading (our mask is > 0), then we cannot take a 0 anymore
                if digit == 0:
                    # if we have already taken a number, now we need to mark the 0 as taken
                    # to prevent numbers like 100 from being counted as special
                    newMask = mask | 1 if mask > 0 else mask
                else:
                    newMask = mask | (1 << digit)

                resForThis += dp(index + 1, newMask, newTight)

            memo[index][mask][tight] = resForThis
            return resForThis

        return dp(0, 0, 1)
